plugins/auto-sizes/includes/improve-calculate-sizes.php

```diff
@@ -135,50 +138,124 @@ function auto_sizes_filter_image_tag( $content, array $parsed_block, WP_Block $b
 /**
  * Modifies the sizes attribute of an image based on layout context.
  *
- * @param int    $id           The image id.
- * @param string $size         The image size data.
- * @param string $align        The image alignment.
- * @param string $resize_width Resize image width.
+ * @since n.e.x.t
+ *
+ * @param int    $id                   The image id.
+ * @param string $size                 The image size data.
+ * @param string $align                The image alignment.
+ * @param string $resize_width         Resize image width.
+ * @param bool   $is_parent_block      Check if image block has parent block.
+ * @param string $ancestor_block_align The ancestor block alignment.
  * @return string The sizes attribute value.
  */
-function auto_sizes_calculate_better_sizes( int $id, string $size, string $align, string $resize_width ): string {
-	$sizes = '';
+function auto_sizes_calculate_better_sizes( int $id, string $size, string $align, string $resize_width, bool $is_parent_block, string $ancestor_block_align ): string {
 	$image = wp_get_attachment_image_src( $id, $size );
 
 	if ( false === $image ) {
-		return $sizes;
+		return '';
 	}
 
 	// Retrieve width from the image tag itself.
 	$image_width = '' !== $resize_width ? (int) $resize_width : $image[1];
 
+	if ( $is_parent_block ) {
+		if ( 'full' === $ancestor_block_align && 'full' === $align ) {
+			return auto_sizes_get_sizes_by_block_alignments( $align, $image_width, true );
+		} elseif ( 'full' !== $ancestor_block_align && 'full' === $align ) {
+			return auto_sizes_get_sizes_by_block_alignments( $ancestor_block_align, $image_width, true );
+		} elseif ( 'full' !== $ancestor_block_align ) {
+			$parent_block_alignment_width = auto_sizes_get_sizes_by_block_alignments( $ancestor_block_align, $image_width );
+			$block_alignment_width        = auto_sizes_get_sizes_by_block_alignments( $align, $image_width );
+			if ( (int) $parent_block_alignment_width < (int) $block_alignment_width ) {
+				return sprintf( '(max-width: %1$s) 100vw, %1$s', $parent_block_alignment_width );
+			} else {
+				return sprintf( '(max-width: %1$s) 100vw, %1$s', $block_alignment_width );
+			}
+		}
+	}
+
+	return auto_sizes_get_sizes_by_block_alignments( $align, $image_width, true );
+}
+
+/**
+ * Generates the `sizes` attribute value based on block information.
+ *
+ * @since n.e.x.t
+ *
+ * @param string $alignment   The alignment.
+ * @param int    $image_width The image width.
+ * @param bool   $print_sizes Print the sizes attribute. Default is false.
+ * @return string The sizes attribute value.
+ */
+function auto_sizes_get_sizes_by_block_alignments( string $alignment, int $image_width, bool $print_sizes = false ): string {
```

Comment:
I don't understand changing the return type based on the `$print_sizes` parameter. If needed, it would be better to break this into two functions if necessary: one that calculates the width value, and another that formats the string for the sizes attribute.

Reply:
Initially, I considered separating them, but since the two functions are similar, I combined them into a single function by adding an extra parameter.

Reply:
+1 to separating them. Using a boolean to change function behavior is a code smell.

Ideally a separate function should take the return value of this function and format it, like @joemcgill suggested.